sw/host/tests/attestation/attestation_test.rs

```diff
@@ -0,0 +1,132 @@
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#![allow(clippy::bool_assert_comparison)]
+use anyhow::{anyhow, Result};
+use base64ct::{Base64, Decoder};
+use clap::Parser;
+use num_bigint_dig::BigUint;
+use regex::Regex;
+use std::time::Duration;
+
+use opentitanlib::app::TransportWrapper;
+use opentitanlib::crypto::sha256;
+use opentitanlib::image::image::Image;
+use opentitanlib::ownership::OwnerBlock;
+use opentitanlib::test_utils::init::InitializeTest;
+use opentitanlib::uart::console::UartConsole;
+use opentitanlib::util::file::FromReader;
+
+use ot_certs::template::{CertificateExtension, Value};
+use ot_certs::x509;
+
+#[derive(Debug, Parser)]
+struct Opts {
+    #[command(flatten)]
+    init: InitializeTest,
+
+    /// Console receive timeout.
+    #[arg(long, value_parser = humantime::parse_duration, default_value = "10s")]
+    timeout: Duration,
+}
+
+// A helper trait for extracting data out of the `Value` type.
+trait GetValue<T> {
+    fn get_value(&self) -> &T;
+}
+
+impl<T: std::fmt::Debug> GetValue<T> for Value<T> {
+    fn get_value(&self) -> &T {
+        match self {
+            Value::Variable(_) => panic!("Not expecting a Variable: {self:?}"),
+            Value::Literal(x) => x,
+        }
+    }
+}
+
+impl<T: std::fmt::Debug> GetValue<T> for Option<Value<T>> {
+    fn get_value(&self) -> &T {
+        match self {
+            None => panic!("Not expecting Option::None"),
+            Some(Value::Variable(_)) => panic!("Not expecting a Variable: {self:?}"),
+            Some(Value::Literal(x)) => x,
+        }
+    }
+}
+
+fn get_base64_blob(haystack: &str, rx: &str) -> Result<Vec<u8>> {
+    let rx = Regex::new(rx)?;
+    let encoded = rx
+        .captures(haystack)
+        .ok_or(anyhow!("Encoded certificate not found"))?;
+    let mut bin = Vec::new();
+    if !encoded[1].is_empty() {
+        let mut decoder = Decoder::<Base64>::new(encoded[1].as_bytes())?;
+        decoder.decode_to_end(&mut bin)?;
+    }
+    Ok(bin)
+}
+
+fn attestation_test(opts: &Opts, transport: &TransportWrapper) -> Result<()> {
+    let uart = transport.uart("console")?;
+    let capture = UartConsole::wait_for(
+        &*uart,
+        r"(?msR)Running.*PASS!$|FAIL!$|BFV:([0-9A-Fa-f]{8})$",
+        opts.timeout,
+    )?;
+
+    let _uds_bin = get_base64_blob(&capture[0], r"(?msR)UDS: (.*?)$")?;
+    let cdi0_bin = get_base64_blob(&capture[0], r"(?msR)CDI_0: (.*?)$")?;
+    let cdi1_bin = get_base64_blob(&capture[0], r"(?msR)CDI_1: (.*?)$")?;
+    let owner_page_0 = get_base64_blob(&capture[0], r"(?msR)OWNER_PAGE_0: (.*?)$")?;
+    let owner_page_1 = get_base64_blob(&capture[0], r"(?msR)OWNER_PAGE_1: (.*?)$")?;
+
+    // TODO: check UDS certificate.
+    let cdi0 = x509::parse_certificate(&cdi0_bin)?;
+    let cdi1 = x509::parse_certificate(&cdi1_bin)?;
+
+    // TODO: verify signature chain from CDI_1 to CDI_0 to UDS.
```

Comment:
This might be a good starting point: https://cs.opensource.google/opentitan/opentitan/+/master:sw/host/provisioning/cert_lib/src/lib.rs;drc=96ea5f430320b411facab840ebe7f785509c4a2a;l=160 but heads up I have not gotten it to work yet with the DICE certs since they have a custom extension that openssl can't seem to parse. Needs further investigation.

Reply:
So openssl won't verify a certificate that it can't fully parse?
One option might be to use the openssl library (not tool) to verify the certificate. From what I remember, it's not that straightforward unfortunately, and it might not have rust bindings, but at least with the library you can parse a certificate and extract extensions and decode manually (which is what I did in the `ot_cert` crate).

Reply:
I'm going to perform the TODOs in follow-on PRs:
- Fix the endianness of the measurements.
- Verify the signature chain.
- Checking the UDS cert on a real chip; skipping UDS on FPGAs.